### ISSUE TITLE
Case-sensitive path handling is error prone on Win/Mac

### ISSUE THREAD

**Issue by sam-mccall** [clangd/clangd#108]
Windows and (most) Mac OS filesystems are case-insensitive but case-preserving.

We use `std::string`/`StringRef` to represent paths for the most part, and do comparisons with `==` and standard string hashing.

This means that if the same path gets into clangd with different case in two places, our comparisons will be incorrect. This affects at least:
 - set of open files (TUScheduler's file map)
 - dynamic index (FileSymbols)
 - background index (shards in memory are keyed by case-sensitive strings, filenames of shards on disk are case-insensitive, but contain case-sensitive hashes of directory paths)
 - global CDB: See #102 for a nasty example.

I think the best option is to have a `Path` class that has the correct semantics for the platform. On e.g. linux this would behave like std::string, on windows it would have case-insensitive ==/hash behavior. Then we get the correct semantics by using Path consistently.

**Comment by sam-mccall** [comment]
This came up again in #665.
And again in https://github.com/ycm-core/YouCompleteMe/issues/3698

One annoyance with the `Path` idea is ownership: paths aren't cheap to copy, so I guess we need a PathRef class too, and the API got a bit out of hand when I tried it.

But this might combine well with an idea I had of interning pathnames in memory, if we can find a sensible way to do it across threads. Then we'd only have PathRefs to deal with.

**Comment by HighCommander4** [comment]
This observation may only be tangentially relevant, but there's precedent for treating paths as distinct types in the [Rust standard library](https://doc.rust-lang.org/std/path/struct.Path.html).

**Comment by sam-mccall** [comment]
Yeah, the details of those types probably don't actually work for us (e.g. C++'s paths are case-sensitive, and the method for checking equivalence does IO, reads through symlinks etc). But the idea of using a proper type is appealing.

https://reviews.llvm.org/D95824 is a sketch of path interning. As noted on that patch it's pretty painful to actually roll out such a change, and I'm wary of getting stuck halfway. So not sure if I want to push for this anytime soon.

(If we got rid of URIs in the index I think it becomes more compelling - having paths be cheap handles is more of a performance win if we're not converting the character data all the time anyway)

**Comment by fifv** [comment]
I met a problem caused by this.

When I rename function names in vscode, it failed with "`Rename failed to apply edits`".
I open the rename preview and it can't open, says "`The editor could not be opened due to an unexpected error: Resource not found in diff editor`".
I have tried diffrent versions of clangd, vscode-clangd extension with no luck.

Then I checked the verbose clangd log, saw the `textDocument/rename` replied with same file twice, one with `D:/` and another one with `d:/`.
Found https://github.com/clangd/clangd/issues/665 and https://github.com/clangd/clangd/issues/1592#issuecomment-1516506136 , I tried to replace all `D:/` in `compile_commands.json` to `d:/`, the problem got fixed.
But `compile_commands.json` will be overriden each time I configure cmake, I didn't find any option about this in cmake docs.

Any hope that this issuewill be resolved? With thanks.

**Comment by narnaud** [comment]
Is there a way to fix that without refactoring everything?
Bug is 5+ years old, it's clear that nobody wants to fix it the best way.

This is impacting all devs using VSCode and clangd plugin on Windows, as they can't use refactoring (and particularly rename) function.

**Comment by HighCommander4** [comment]
There is a [workaround](https://github.com/clangd/vscode-clangd/issues/687#issuecomment-2365439284) that came up in one of the linked issues, copying it here for visibility:

> What I'm seeing in the log, is that vscode is sending a lowercase `c:` in the path in the `textDocument/didOpen` command. But the `compile_commands.json` file uses uppercase `C:`.
>
> Since you probably can't control what vscode sends, I would try changing the `compile_commands.json` to contain lowercase `c:`'s instead.

In other words, ensure that the paths in `compile_commands.json` match the case of the paths vscode sends over LSP, either by getting the tool that generates `compile_commands.json` to produce that in the first place, or by running a build step that post-processes the `compile_commands.json` to adjust the paths.

As for whether the underlying issue has a fix that doesn't involve a lot of refactoring, I'm not sure. At least nothing obvious comes to mind. I'm happy to help evaluate alternative proposals if someone has one.

**Comment by narnaud** [comment]
> In other words, ensure that the paths in `compile_commands.json` match the case of the paths vscode sends over LSP, either by getting the tool that generates `compile_commands.json` to produce that in the first place, or by running a build step that post-processes the `compile_commands.json` to adjust the paths.

I gave it a try at CMake, unfortunately it doesn't look like it's going to change:
https://gitlab.kitware.com/cmake/cmake/-/issues/26860#note_1647726

Having a post-build step is also not desirable, as it will be fixed for one project, but you need to add it to all other projects (and it takes compile time for nothing for those not using the file).

**Comment by timxx** [comment]
I have also encountered this problem myself, and it significantly affects my experience using clangd on Windows.

Is there any plan or timeline for addressing this issue?

**Comment by HighCommander4** [comment]
> Is there any plan or timeline for addressing this issue?

There is work in progress towards a fix at https://github.com/llvm/llvm-project/pull/136439.

Meanwhile, the workaround I recommend is to normalize the paths in `compile_commands.json` in a post-build step to be consistent with what the editor sends (e.g. to lowercase for vscode).

**Comment by listenerri** [comment]
I created a bat script for this problem: https://github.com/listenerri/scripts/blob/master/bat/cmake-fix-clangd.bat, which is used to wrap the cmake command. After the real cmake command is executed, all uppercase letters in compile_commands.json are changed to lowercase.

Note: The script uses the sed command of msys2 to perform the modification operation. If you don't install msys2 suite, you need to modify the script to perform the modification operation in other ways.

Usage: put the absolute path of this script in the path environment variable, and then modify the config of vscode, change the path of cmake to: cmake-fix-clangd.bat.

```
# vscode settings.json
"cmake.cmakePath": "cmake-fix-clangd.bat"
```

**Comment by psycha0s** [comment]
I tried the [vscode-clangd patch](https://github.com/clangd/vscode-clangd/pull/747), and it worked for me, but only partially - I still had issues with header/source switching. I tried [the .bat script](https://github.com/clangd/clangd/issues/108#issuecomment-3587862830), but it has issues with additional CMake arguments that cmake-tools passes through on the command line (I set up some debug flags in settings.ini):
```json
"cmake.configureSettings": { "CMAKE_CXX_FLAGS_DEBUG": "-g3 -O0 -Wall -Wextra -Wpedantic -fno-omit-frame-pointer" },
```
Despite strongly agreeing with the CMake folks that clangd shouldn't compare paths by string matching, I think the easiest workaround right now is patching CMake itself. I created [a patch for CMake 4.1.2 to lowercase all drive letters in compile_commands.json](https://gist.github.com/psycha0s/61db38917f0ba2dc7d04173cd04a09af) on Windows

I hope this helps someone and that this issue will finally be resolved in clangd.
